File: CORE/src/main/java/com/variant/core/event/EventWriter.java
package com.variant.core.event;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.concurrent.ConcurrentLinkedQueue;

import org.apache.commons.lang3.time.DurationFormatUtils;
import org.slf4j.Logger;

import com.variant.core.Variant;
import com.variant.core.VariantInternalException;
import com.variant.core.VariantProperties;
import com.variant.core.impl.VariantCoreImpl;

public class EventWriter {
	
	// The underlying buffer is a non-blocking, unbounded queue. We will enforce the soft upper bound,
	// refusing inserts that will put the queue size over the limit, but not worrying about
	// a possible overage due to concurrency.
	private ConcurrentLinkedQueue<VariantEventSupport> eventQueue = null;

	// Max queue size (soft).
	private int queueSize;
	
	// Number of entries in the queue that will trigger the wakeup of the persister thread.
	private int pctFullSize;
	
	// Number of entries in the queue that we won't attempt to get under.
	private int pctEmptySize;
	
	// The persister thread will wake up at least this frequently and flush the queue.
	private long maxPersisterDelayMillis;
		
	// Asynchronous persister thread consumes events from the holding queue.
	private Thread persisterThread;
	
	// The actual event persister passed to the constructor by client code.
	private EventPersister persisterImpl = null;
	
	// Logger
	private Logger logger = ((VariantCoreImpl)Variant.Factory.getInstance()).getLogger();
	
	/**
	 * Validate that event is well-formed.
	 * @param event
	 */
	private void validateEvent(VariantEventSupport event) {
		if (event.experiences.size() == 0) 
			throw new VariantInternalException(
					String.format("Event [%s] [%s] has no experiences. Ignored.", event.eventName, event.eventValue));
	}
	
	/**
	 * Expose event persister to tests via package visibility.
	 * @return
	 */
	public EventPersister getEventPersister() {
		return persisterImpl;
	}
	
	//---------------------------------------------------------------------------------------------//
	//                                          PUBLIC                                             //
	//---------------------------------------------------------------------------------------------//
	
	/**
	 * Constructor
	 */
	public EventWriter(EventPersister persisterImpl) {
		
		this.persisterImpl = persisterImpl;
		this.queueSize = VariantProperties.getInstance().eventWriterBufferSize();
		this.pctFullSize = queueSize * VariantProperties.getInstance().eventWriterPercentFull() / 100;
		this.pctEmptySize = (int) Math.ceil(queueSize * 0.1);
		this.maxPersisterDelayMillis = VariantProperties.getInstance().eventWriterMaxDelayMillis();
		
		eventQueue = new ConcurrentLinkedQueue<VariantEventSupport>();
		
		persisterThread = new Thread(new PersisterThread());
		
		// Not a daemon: intercept interrupt and flush the buffer before exiting.
		persisterThread.setDaemon(false);
		persisterThread.setName(PersisterThread.class.getSimpleName());
		persisterThread.start();
	}
		
	/**
	 * Shutdown this event writer.
	 * Cannot be used after this.
	 */
	public void shutdown() {
		long now = System.currentTimeMillis();
		persisterThread.interrupt();
		persisterThread = null;
		if (logger.isDebugEnabled()) {
			logger.debug(
					"Event Writer shutdown in " + (DurationFormatUtils.formatDuration(System.currentTimeMillis() - now, "mm:ss.SSS")));
		}
	}
	
	/**
	 * Write collection of events to the queue.  This method never blocks:
	 * if there's no room on the queue to hold all the events, write as many as we can in the
	 * order of the collection's iterator and ignore the rest, returning back to the caller 
	 * the number of elements that were accepted. The caller may re-attempt, drop, log, etc.
	 *  
	 * @param events
	 * @return number of elements actually written.
	 */
	public int write(Collection<VariantEventSupport> events) {
		
		// size() is an O(n) operation - do it once.
		int currentQueueSize = eventQueue.size();
		int acceptCount = 0;
		
		Iterator<VariantEventSupport> iter = events.iterator();
		while (currentQueueSize < queueSize && iter.hasNext()) {
			VariantEventSupport event = iter.next();
			validateEvent(event);
			eventQueue.add(event);
			currentQueueSize++;
			acceptCount++;
		}
		
		// Block momentarily to wake up the persister thread if the queue has reached the pctFull size.
		synchronized (eventQueue) {
			if (currentQueueSize >= pctFullSize) eventQueue.notify();
		}

		return acceptCount;
	}
		
	/**
	 * Enqueue a single event.  Variation of the above.
	 *  
	 * @param event
	 */
	public void write(VariantEventSupport event) {
		ArrayList<VariantEventSupport> collection = new ArrayList<VariantEventSupport>(1);
		collection.add(event);
		write(collection);
	}
	
	/**
	 * Persister thread.
	 * Removes events from the queue and flushes them to an event persistence interface. 
	 * 
	 * @author Igor.
	 *
	 */
	private class PersisterThread implements Runnable {
		
		@Override
		public void run() {

			logger.debug("Event persister thread " + Thread.currentThread().getName() + " created.");
			
			boolean InterruptedExceptionThrown = false;
			
			while (true) {
				
				try {
					// We were either woken up because the queue is over pctFull,
					// or we timed out waiting. Flush either way, event if under pctEmpty,
					// and the keep flushing until under pctEmpty.
					do {
						flush();
					} while (eventQueue.size() >= pctEmptySize);
										
					// Block until the queue is over pctFull again, but with timeout.
					synchronized (eventQueue) {
						eventQueue.wait(maxPersisterDelayMillis);
					}

				}
				catch (InterruptedException e) {
					InterruptedExceptionThrown = true;
				}
				catch (Throwable t) {
					logger.error("Unexpected exception in async database event writer.", t);
				}
				
				if (InterruptedExceptionThrown || Thread.currentThread().isInterrupted()) {
					try {
						flush();
					}
					catch (Throwable t) {
						logger.error("Unexpected exception in async database event writer.", t);
					}
					logger.debug("Persister thread " + Thread.currentThread().getName() + " interrupted and exited.");
					return;
				};
			}
			
		}		
		
		/**
		 * Flush the entire queue to an event persister.
		 */
		private void flush() throws Exception {

			ArrayList<VariantEventSupport> events = new ArrayList<VariantEventSupport>();

			VariantEventSupport event = eventQueue.poll();
			while (event != null) {
				events.add(event);
				event = eventQueue.poll();
			}

			if (events.isEmpty()) return;
			
			long now = System.currentTimeMillis();
			persisterImpl.persist(events);		
			logger.debug("Wrote " + events.size() + " events in " + DurationFormatUtils.formatDurationHMS(System.currentTimeMillis() - now));
		}
	}
}

File: SERVER/src/main/scala/com/variant/server/trace/EventWriter.scala
package com.variant.server.impl
/*
import java.util.LinkedList
import java.util.concurrent.ConcurrentLinkedQueue
import com.variant.server.api.FlushableTraceEvent
import com.variant.server.boot.VariantServer
import com.variant.server.schema.ServerFlusherService
import java.time.Instant
import java.time.Duration
import com.variant.core.util.TimeUtils
import com.typesafe.scalalogging.LazyLogging
import akka.event.jul.Logger

class TraceEventWriter(private val flushService: ServerFlusherService)(implicit server: VariantServer)
   extends LazyLogging with ConfigKeys {

   private val config = server.config

   val flusher = flushService.getFlusher

   // The underlying buffer is a non-blocking, unbounded queue. We will enforce the soft upper bound,
   // refusing inserts that will put the queue size over the limit, but not worrying about
   // a possible overage due to concurrency.
   private val bufferQueue = new ConcurrentLinkedQueue[FlushableTraceEvent]()

   // The flusher thread.
   private val flusherThread = new FlusherThread();
   // Not a daemon. Intercept interrupt and flush the buffer before exiting.
   flusherThread.setDaemon(false)
   flusherThread.setName("Event Flusher For Schema " + flushService.getSchema.getMeta.getName)
   flusherThread.start()

   //---------------------------------------------------------------------------------------------//
   //                                          PUBLIC                                             //
   //---------------------------------------------------------------------------------------------//

   logger.debug("Event writer started.");

   /**
    * Flush this writer immediately, without waiting for the internal flush.
    */
   def flush() = {
      bufferQueue.synchronized {
         flusherThread.flush()
      }
   }

   /**
    * Shutdown this event writer.
    * Cannot be used after this.
    */
   def shutdown() {
      // Destroy the flusher thread.
      // TODO this should be an actor
      flusherThread.interrupt()

      // One last flush
      flush()

      flusher.destroy()

      logger.debug("Event writer shutdown.");
   }

   /**
    * Write collection of events to the queue.  This method never blocks:
    * if there's no room on the queue to hold all the events, write as many as we can in the
    * order of the collection's iterator and ignore the rest. Log ERROR dropped events.
    *
    * @param event decoratedEvent
    *
    * @return number of elements actually written.
    */
   def write(event: FlushableTraceEvent) {

      // We don't worry about possible concurrent writes because the underlying
      // queue implementation is thread safe and unbound.  It's okay to temporarily
      // go over the maxQueueSize due to concurrency, so long as we eventually shrink back.
      // But we won't go over it knowingly.
      val currentSize = bufferQueue.size();
      var dropCount = 0

      if (currentSize < maxBufferSize) {
         bufferQueue.add(event);
      } else {
         val msg = "Dropped %d event(s) due to buffer overflow. Consider increasing %s system property (current value %d)"
         logger.trace(msg.format(1, EVENT_WRITER_BUFFER_SIZE, maxBufferSize))
         dropCount += 1
         if (dropCount % 1000 == 0)
            logger.info(msg.format(1000, EVENT_WRITER_BUFFER_SIZE, maxBufferSize))
      }

      // Block momentarily to wake up the flusher thread if the queue has reached the full size.
      bufferQueue.synchronized {
         if (currentSize >= fullSize) bufferQueue.notify();
      }

   }

   /**
    * Flusher thread.
    * Removes events from the queue and flushes them to an event persistence interface.
    *
    * @author Igor.
    *
    */
   class FlusherThread extends Thread {

      override def run() {

         var interruptedExceptionThrown = false;
         var timeToGo = false;
         while (!timeToGo) {

            try {
               // We were either woken up because the queue is over pctFull,
               // or we timed out waiting. Flush either way, event if under pctEmpty,
               // and the keep flushing until under pctEmpty.
               do {
                  flush();
               } while (!isInterrupted() && bufferQueue.size() >= emptySize);

               // Block until the queue is over pctFull again, but with timeout.
               if (!isInterrupted()) {
                  bufferQueue.synchronized {
                     bufferQueue.wait(maxDelayMillis);
                  }
               }
            } catch {
               case _: InterruptedException => interruptedExceptionThrown = true
               case t: Throwable => {
                  logger.error("bufferQueue = " + bufferQueue);
                  logger.error("Unhandled exception in event writer", t)
               }
            }

            if (interruptedExceptionThrown || isInterrupted()) {
               try {
                  flush();
               } catch {
                  case t: Throwable => logger.error("Unhandled exception in event flusher", t);
               }
               timeToGo = true
            }
         }
      }

      /**
       * Flush the entire buffer.
       */
      def flush() {

         var events = new LinkedList[FlushableTraceEvent]();
         while (bufferQueue != null && !bufferQueue.isEmpty()) events.add(bufferQueue.poll());
         logger.debug(s"About to flush ${events.size} events")
         if (!events.isEmpty()) {
            var start = Instant.now();
            flusher.flush(events);
            logger.info("Flushed " + events.size() + " event(s) in " + TimeUtils.formatDuration(Duration.between(start, Instant.now())))
         }
      }
   }

}
*/ 